### ISSUE TITLE
Operation timed out after 30000 milliseconds with 0 bytes received

### ISSUE THREAD

**Issue by toanit247** [facebook/php-webdriver#598]
Hi,
I received a notice that "Operation timed out after 30000 milliseconds with 0 bytes received."
Maybe somebody can help me to caught this event, i really need to close my Chrome Browser.
We hope to receive the help from you guys about this issue.

Best regards,

**Comment by techsemicolon** [comment]
I am having this issue also :/

**Comment by jeremy-smith-maco** [comment]
You should probably provide like a stack trace, some test code or something more useful as no one can really help you with the information you have provided.

**Comment by mb0000** [comment]
Is this the same symptom (and possible cause(s)) as https://github.com/facebook/php-webdriver/issues/520?

**Comment by porjo** [comment]
I was getting this error using a hub/node (grid) setup using Docker containers (from [SeleniumHQ/docker-selenium](https://github.com/SeleniumHQ/docker-selenium)). On a hunch I added more RAM to the host and the errors stopped. In my case at least, it would seem that there was insufficient RAM for the number of nodes I was attempting to run simultaneously (even though I wasn't seeing any errors to indicate this in the console output of the nodes). Maybe try throwing more RAM at it?

**Comment by denvit** [comment]
I was having the same problem with Laravel Dusk. I've solved the problem when turned off xdebug in PHPStorm.

**Comment by vedvikramsinghthakur** [comment]
I found the solution.
Use this code:
$driver=RemoteWebDriver::create($host,$capabilities,3600000,3600000);
Or you can take refrence from-
https://facebook.github.io/php-webdriver/community/Facebook/WebDriver/Remote/RemoteWebDriver.html

**Comment by OndraM** [comment]
Hi, I'm closing this for now because there hasn't been any new information for a while and the issue might be already resolved. 

But if the issue still persist, please don't hesitate to open a new one. However please fill the full issue template which is provided when creating new issue (including code you run and full error log from selenium/chromedriver etc.) - without these information we might not be able to help you.

**Comment by andrewnicols** [comment]
As it happens I’ve been trying to chase some of these issues down too.

I believe it’s a bug with either gecko driver or Firefox in my case and it’s intermittent. There were some issues with Firefox headless on Firefox 71.0 but they’re fixed in 73.0.

**Comment by OndraM** [comment]
It may be anything - but without any code sample and error log we can't know :). 

There were few reports about issues like this, however none of them was reproducible so far :disappointed: .

**Comment by cerw** [comment]
This is still a problem. random one but it is.

**Comment by andrewnicols** [comment]
I’ve investigated this and similar issues and none of them have been a problem with this driver.

There are several things at play:

So if you’re using Selenium, when you request a new session Selenium farms that out to a grid node, which starts the relevant driver (i.e. chromedriver), and requests a session. You get a new sessionid, and a browser window. After that point all of your URLs include that sessionid (i.e. something like http://localhost:4444/wd/hub/[sessionid]/element from memory), and Selenium is basically acting as a proxy and relaying your traffic to the relevant driver.

Once it gets to the driver, the [W3C specification defines a series of timeouts](https://w3c.github.io/webdriver/#timeouts). Broadly these are the `script` timeout (/execute/sync and /execute/async), the `pageLoad` timeout (anything causing page navigation), and the `implicit` timeout (searching for elements). These default to 30 seconds, 300 seconds, and no limit by default.

And then you have the browser itself, which can be limited by a range of things.

So looking at those three layers:
- your application
- php-webdriver
- [selenium] (if used)
- driver (chromedriver/geckodriver/etc.)
- browser

The first time to look at is if you’re using Selenium. From Selenium 4.0 onwards (possibly Selenium 3.141.59 too but I haven’t seen this there and have not checked the code), there is a hard-coded 30 second timeout on all requests proxied to the remote end. In Selenium 3 this can be controlled with the `--timeout` flag when starting Selenium, but this does not exist in Selenium 4. See this issue for more info: https://github.com/SeleniumHQ/selenium/issues/9045
What this means is that if you make a request, for example “execute this script” or “click this button which loads a new page” and the result takes more than 30 seconds to complete (slow script, or slow page), Selenium will timeout and you will see this error. The error is _reported_ by the Curl call that we make in php-webdriver because it’s just been disconnected from the Selenium server with no response at all. It’s literally saying that it received no data after 30,000ms. We have no more information to give.
If you use Selenium *this is your most likely problem*.

We then get to the driver itself - that could be Chromedriver, Geckodriver, Edgedriver, etc. Now these guys have a timeout too - that’s what’s set in the specification. However, if you hit any of those three timeouts (script, pageLoad, implicit) you _should_ get an error message to tell you that your request timed out. That request will come from the relevant driver, and it will _not_ be this 0 bytes received message.
In addition you can already configure that timeout using the setTimeout call. We already support it in php-webdriver and you just need to call it.
This is _unlikely_ to be the cause of this problem because the error message would be different.

The final point is the browser itself. This one gets a bit weird because there’s a lot more going on, but with Chrome this issue is sometimes seen when the browser is doing GPU- based processing, or when you’re running low on memory. I would generally recommend that you run with the `no-gpu` argument when starting the session as this disables GPU-based processing. The other argument to be aware of is `no-sandbox` - I forget the detail of this one but many report that it has helped them in the same situation. Finally, if you run Selenium/Chrome in a docker image then please make sure that you are mapping shared memory to the container. You can do this with the `--shm-max` flag to the `docker run` command or by mapping `/dev/shm` between host and container.

On top of this I’ve seen reports on chromedriver with similar timeouts but I haven’t seen any resolution to these, nor any confirmation that they actually represent a bug.

So basically:
- this is almost certainly _not_ a bug in php-webdriver;
- it probably actually _is_ a slow call;
- you are probably using Selenium and need to bump the timeout;
- but it could be an issue with how you run Chrome, either:
- - when you start the session; or
- - the way in which you run Chrome inside a container or similar.

**Comment by duha000750** [comment]
> I found the solution.
> Use this code:
> $driver=RemoteWebDriver::create($host,$capabilities,3600000,3600000);
> Or you can take refrence from-
> https://facebook.github.io/php-webdriver/community/Facebook/WebDriver/Remote/RemoteWebDriver.html

thank you very much!!

**Comment by robertoitec** [comment]
> $driver=RemoteWebDriver::create($host,$capabilities,3600000,3600000);

thank you very much!

**Comment by PavelFil** [comment]
The problem is anti ddons protection systems. These systems can drop connection (not reject, just drop connection) by some reason. Browser like Chrome will be wait for answer for 30 seconds. After that he drops stuck connection and ChromeDriver can answer to the program.

So, because of default Chrome Browser file load timeout of 30 seconds, any your selenium timeout (implicitly, pageLoad, script execution, request) should be above 30 seconds.
